parsing/tools/src/main/java/com/apache/calcite/parsing/tools/FindDialectParsingErrors.java

```diff
@@ -0,0 +1,424 @@
+import org.apache.calcite.runtime.Resources;
+import org.apache.calcite.sql.parser.SqlParseException;
+import org.apache.calcite.sql.parser.SqlParser;
+import org.apache.calcite.sql.parser.SqlParserImplFactory;
+import org.apache.calcite.sql.parser.bigquery.BigQueryParserImpl;
+import org.apache.calcite.sql.parser.defaultdialect.DefaultDialectParserImpl;
+import org.apache.calcite.sql.parser.dialect1.Dialect1ParserImpl;
+import org.apache.calcite.sql.parser.hive.HiveParserImpl;
+import org.apache.calcite.sql.parser.mysql.MySQLParserImpl;
+import org.apache.calcite.sql.parser.postgresql.PostgreSQLParserImpl;
+import org.apache.calcite.sql.parser.redshift.RedshiftParserImpl;
+
+import au.com.bytecode.opencsv.CSVReader;
+
+import com.google.gson.stream.JsonWriter;
+
+import java.io.BufferedWriter;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.OutputStreamWriter;
+import java.lang.reflect.Method;
+import java.nio.charset.StandardCharsets;
+import java.text.MessageFormat;
+import java.text.ParseException;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * Processes a CSV file containing queries and will output a JSON file
+ * containing the results of failing queries.
+ */
+public class FindDialectParsingErrors {
+
+  private final String inputPath;
+  private final String outputPath;
+  private final Dialect dialect;
+  private final boolean groupByErrors;
+  private final int numSampleQueries;
+  private final List<MessageFormat> errorFormats;
+
+  /**
+   *  Creates a new instance of {@code FindDialectParsingErrors}. Also populates
+   *  the errorFormats list with a {@code MessageFormat} object for each custom
+   *  defined error message in CalciteResource.
+   *
+   * @param inputPath Path to the input CSV file containing queries
+   * @param outputPath Path to the output JSON file containing results of
+   *                   failing queries
+   * @param dialect Specifies which dialectic parser to use for processing
+   *                queries
+   * @param groupByErrors Specifies if the output format should group failing
+   *                      queries by error message type
+   * @param numSampleQueries The max number of sample queries to show for an
+   *                         error type when grouping queries by error message
+   *                         type
+   */
+  public FindDialectParsingErrors(String inputPath, String outputPath,
+      Dialect dialect, boolean groupByErrors, int numSampleQueries) {
+    this.inputPath = inputPath;
+    this.outputPath = outputPath;
+    this.dialect = dialect;
+    this.groupByErrors = groupByErrors;
+    this.numSampleQueries = numSampleQueries;
+    errorFormats = new ArrayList<>();
```

Comment:
should this be this.errorFormats?

Reply:
I believe it's legal for it to be with or without the "this" prefix. I'm guessing Michael didn't use the "this" prefix in this case because it wasn't needed to disambiguate it from a local variable with the same name.

Nevertheless, it might just be clearer to use "this.errorFormats" in the constructor (but not necessarily anywhere else) since the other fields are being set this way.

Reply:
Sure I can add that in for clarification